### PR TITLE
キャラの強さを判定するクラスを作成

### PR DESCRIPTION
**Pull request by yumayo14**
## やったこと
- 渡された複数のキャラクター達の中で、「攻撃力と防御力の合計値が最も高いキャラクター達」、「攻撃力が最も高いキャラクター達」、「防御力が最も高いキャラクター達」の配列を取得できるようにした。

### merge先について
マージ先がmasterブランチでない理由
- add_guild_interfaceとmasterの変更差分を取りこまない様にし、レビューをしやすくするため。

最終的なマージ先
- add_guildr_interfaceがマージされた後、pull --rebaseで変更差分を取り込み、masterへとマージする。

### pry上での出力結果
- 必要なファイルのインポート
[1] pry(main)> require_relative 'character_checker.rb'
=> true
[2] pry(main)> require_relative 'guild.rb'
=> true
- `CharacterCheckerのインスタンスを作成`
[3] pry(main)> character_checker = CharacterChecker.new(Guild.new.member)
=> #<CharacterChecker:0x00007ff23f844638
 @characters=
  [#<Fighter:0x00007ff23e44e278
    @element=#<Wind:0x00007ff23e44e390 @name="wind">,
    @equipment=#<Sword:0x00007ff23e44e318 @name="sword">,
    @sex=#<Man:0x00007ff23e44e408 @name="man">>,
   #<Fighter:0x00007ff23e44e0e8
    @element=#<Wind:0x00007ff23e44e200 @name="wind">,
    @equipment=#<Stick:0x00007ff23e44e188 @name="stick">,
    @sex=#<Man:0x00007ff23e44e250 @name="man">>,
   #<Fighter:0x00007ff23e44df80
    @element=#<Wind:0x00007ff23e44e070 @name="wind">,
    @equipment=#<Glove:0x00007ff23e44e020 @name="glove">,
    @sex=#<Man:0x00007ff23e44e0c0 @name="man">>,
   #<Fighter:0x00007ff23e44dda0
    @element=#<Water:0x00007ff23e44deb8 @name="water">,
    @equipment=#<Sword:0x00007ff23e44de68 @name="sword">,
    @sex=#<Man:0x00007ff23e44df08 @name="man">>,
   #<Fighter:0x00007ff23e44dc10
    @element=#<Water:0x00007ff23e44dd00 @name="water">,
    @equipment=#<Stick:0x00007ff23e44dc88 @name="stick">,
    @sex=#<Man:0x00007ff23e44dd78 @name="man">>,
   #<Fighter:0x00007ff23e44daa8
    @element=#<Water:0x00007ff23e44db98 @name="water">,
    @equipment=#<Glove:0x00007ff23e44db48 @name="glove">,
    @sex=#<Man:0x00007ff23e44dbe8 @name="man">>,
   #<Fighter:0x00007ff23e44d058
    @element=#<Thunder:0x00007ff23e44d198 @name="thunder">,
    @equipment=#<Sword:0x00007ff23e44d120 @name="sword">,
    @sex=#<Man:0x00007ff23e44d9b8 @name="man">>,
   #<Fighter:0x00007ff23e44cb58
    @element=#<Thunder:0x00007ff23e44cdb0 @name="thunder">,
    @equipment=#<Stick:0x00007ff23e44ccc0 @name="stick">,
    @sex=#<Man:0x00007ff23e44d008 @name="man">>,
   #<Fighter:0x00007ff23e44c8b0
    @element=#<Thunder:0x00007ff23e44ca40 @name="thunder">,
    @equipment=#<Glove:0x00007ff23e44c9a0 @name="glove">,
    @sex=#<Man:0x00007ff23e44cb08 @name="man">>,
   #<Fighter:0x00007ff23e44c388
    @element=#<Wind:0x00007ff23e44c540 @name="wind">,
    @equipment=#<Sword:0x00007ff23e44c400 @name="sword">,
    @sex=#<Woman:0x00007ff23e44c658 @name="woman">>,
   #<Fighter:0x00007ff23e44c040
    @element=#<Wind:0x00007ff23e44c270 @name="wind">,
    @equipment=#<Stick:0x00007ff23e44c1a8 @name="stick">,
    @sex=#<Woman:0x00007ff23e44c2c0 @name="woman">>,
   #<Fighter:0x00007ff23f847ea0
    @element=#<Wind:0x00007ff23f847f68 @name="wind">,
    @equipment=#<Glove:0x00007ff23f847f18 @name="glove">,
    @sex=#<Woman:0x00007ff23e44c018 @name="woman">>,
   #<Fighter:0x00007ff23f847d38
    @element=#<Water:0x00007ff23f847e00 @name="water">,
    @equipment=#<Sword:0x00007ff23f847db0 @name="sword">,
    @sex=#<Woman:0x00007ff23f847e50 @name="woman">>,
   #<Fighter:0x00007ff23f847bf8
    @element=#<Water:0x00007ff23f847cc0 @name="water">,
    @equipment=#<Stick:0x00007ff23f847c70 @name="stick">,
    @sex=#<Woman:0x00007ff23f847d10 @name="woman">>,
   #<Fighter:0x00007ff23f847ab8
    @element=#<Water:0x00007ff23f847b80 @name="water">,
    @equipment=#<Glove:0x00007ff23f847b30 @name="glove">,
    @sex=#<Woman:0x00007ff23f847bd0 @name="woman">>,
   #<Fighter:0x00007ff23f847950
    @element=#<Thunder:0x00007ff23f847a18 @name="thunder">,
    @equipment=#<Sword:0x00007ff23f8479c8 @name="sword">,
    @sex=#<Woman:0x00007ff23f847a68 @name="woman">>,
   #<Fighter:0x00007ff23f847810
    @element=#<Thunder:0x00007ff23f8478d8 @name="thunder">,
    @equipment=#<Stick:0x00007ff23f847888 @name="stick">,
    @sex=#<Woman:0x00007ff23f847928 @name="woman">>,
   #<Fighter:0x00007ff23f8476d0
    @element=#<Thunder:0x00007ff23f847798 @name="thunder">,
    @equipment=#<Glove:0x00007ff23f847748 @name="glove">,
    @sex=#<Woman:0x00007ff23f8477e8 @name="woman">>,
   #<Warrior:0x00007ff23f8474f0
    @element=#<Wind:0x00007ff23f8475b8 @name="wind">,
    @equipment=#<Sword:0x00007ff23f847568 @name="sword">,
    @sex=#<Man:0x00007ff23f847608 @name="man">>,
   #<Warrior:0x00007ff23f8473b0
    @element=#<Wind:0x00007ff23f847478 @name="wind">,
    @equipment=#<Stick:0x00007ff23f847428 @name="stick">,
    @sex=#<Man:0x00007ff23f8474c8 @name="man">>,
   #<Warrior:0x00007ff23f847270
    @element=#<Wind:0x00007ff23f847338 @name="wind">,
    @equipment=#<Glove:0x00007ff23f8472e8 @name="glove">,
    @sex=#<Man:0x00007ff23f847388 @name="man">>,
   #<Warrior:0x00007ff23f847108
    @element=#<Water:0x00007ff23f8471d0 @name="water">,
    @equipment=#<Sword:0x00007ff23f847180 @name="sword">,
    @sex=#<Man:0x00007ff23f847220 @name="man">>,
   #<Warrior:0x00007ff23f846fc8
    @element=#<Water:0x00007ff23f847090 @name="water">,
    @equipment=#<Stick:0x00007ff23f847040 @name="stick">,
    @sex=#<Man:0x00007ff23f8470e0 @name="man">>,
   #<Warrior:0x00007ff23f846e88
    @element=#<Water:0x00007ff23f846f50 @name="water">,
    @equipment=#<Glove:0x00007ff23f846f00 @name="glove">,
    @sex=#<Man:0x00007ff23f846fa0 @name="man">>,
   #<Warrior:0x00007ff23f846d20
    @element=#<Thunder:0x00007ff23f846de8 @name="thunder">,
    @equipment=#<Sword:0x00007ff23f846d98 @name="sword">,
    @sex=#<Man:0x00007ff23f846e38 @name="man">>,
   #<Warrior:0x00007ff23f846be0
    @element=#<Thunder:0x00007ff23f846ca8 @name="thunder">,
    @equipment=#<Stick:0x00007ff23f846c58 @name="stick">,
    @sex=#<Man:0x00007ff23f846cf8 @name="man">>,
   #<Warrior:0x00007ff23f846aa0
    @element=#<Thunder:0x00007ff23f846b68 @name="thunder">,
    @equipment=#<Glove:0x00007ff23f846b18 @name="glove">,
    @sex=#<Man:0x00007ff23f846bb8 @name="man">>,
   #<Warrior:0x00007ff23f846910
    @element=#<Wind:0x00007ff23f8469d8 @name="wind">,
    @equipment=#<Sword:0x00007ff23f846988 @name="sword">,
    @sex=#<Woman:0x00007ff23f846a28 @name="woman">>,
   #<Warrior:0x00007ff23f8467d0
    @element=#<Wind:0x00007ff23f846898 @name="wind">,
    @equipment=#<Stick:0x00007ff23f846848 @name="stick">,
    @sex=#<Woman:0x00007ff23f8468e8 @name="woman">>,
   #<Warrior:0x00007ff23f846690
    @element=#<Wind:0x00007ff23f846758 @name="wind">,
    @equipment=#<Glove:0x00007ff23f846708 @name="glove">,
    @sex=#<Woman:0x00007ff23f8467a8 @name="woman">>,
   #<Warrior:0x00007ff23f846528
    @element=#<Water:0x00007ff23f8465f0 @name="water">,
    @equipment=#<Sword:0x00007ff23f8465a0 @name="sword">,
    @sex=#<Woman:0x00007ff23f846640 @name="woman">>,
   #<Warrior:0x00007ff23f8463e8
    @element=#<Water:0x00007ff23f8464b0 @name="water">,
    @equipment=#<Stick:0x00007ff23f846460 @name="stick">,
    @sex=#<Woman:0x00007ff23f846500 @name="woman">>,
   #<Warrior:0x00007ff23f8462a8
    @element=#<Water:0x00007ff23f846370 @name="water">,
    @equipment=#<Glove:0x00007ff23f846320 @name="glove">,
    @sex=#<Woman:0x00007ff23f8463c0 @name="woman">>,
   #<Warrior:0x00007ff23f846140
    @element=#<Thunder:0x00007ff23f846208 @name="thunder">,
    @equipment=#<Sword:0x00007ff23f8461b8 @name="sword">,
    @sex=#<Woman:0x00007ff23f846258 @name="woman">>,
   #<Warrior:0x00007ff23f846000
    @element=#<Thunder:0x00007ff23f8460c8 @name="thunder">,
    @equipment=#<Stick:0x00007ff23f846078 @name="stick">,
    @sex=#<Woman:0x00007ff23f846118 @name="woman">>,
   #<Warrior:0x00007ff23f845ec0
    @element=#<Thunder:0x00007ff23f845f88 @name="thunder">,
    @equipment=#<Glove:0x00007ff23f845f38 @name="glove">,
    @sex=#<Woman:0x00007ff23f845fd8 @name="woman">>,
   #<Wizard:0x00007ff23f845cb8
    @element=#<Wind:0x00007ff23f845d80 @name="wind">,
    @equipment=#<Sword:0x00007ff23f845d30 @name="sword">,
    @sex=#<Man:0x00007ff23f845dd0 @name="man">>,
   #<Wizard:0x00007ff23f845b78
    @element=#<Wind:0x00007ff23f845c40 @name="wind">,
    @equipment=#<Stick:0x00007ff23f845bf0 @name="stick">,
    @sex=#<Man:0x00007ff23f845c90 @name="man">>,
   #<Wizard:0x00007ff23f845a38
    @element=#<Wind:0x00007ff23f845b00 @name="wind">,
    @equipment=#<Glove:0x00007ff23f845ab0 @name="glove">,
    @sex=#<Man:0x00007ff23f845b50 @name="man">>,
   #<Wizard:0x00007ff23f8458d0
    @element=#<Water:0x00007ff23f845998 @name="water">,
    @equipment=#<Sword:0x00007ff23f845948 @name="sword">,
    @sex=#<Man:0x00007ff23f8459e8 @name="man">>,
   #<Wizard:0x00007ff23f845790
    @element=#<Water:0x00007ff23f845858 @name="water">,
    @equipment=#<Stick:0x00007ff23f845808 @name="stick">,
    @sex=#<Man:0x00007ff23f8458a8 @name="man">>,
   #<Wizard:0x00007ff23f845650
    @element=#<Water:0x00007ff23f845718 @name="water">,
    @equipment=#<Glove:0x00007ff23f8456c8 @name="glove">,
    @sex=#<Man:0x00007ff23f845768 @name="man">>,
   #<Wizard:0x00007ff23f8454e8
    @element=#<Thunder:0x00007ff23f8455b0 @name="thunder">,
    @equipment=#<Sword:0x00007ff23f845560 @name="sword">,
    @sex=#<Man:0x00007ff23f845600 @name="man">>,
   #<Wizard:0x00007ff23f8453a8
    @element=#<Thunder:0x00007ff23f845470 @name="thunder">,
    @equipment=#<Stick:0x00007ff23f845420 @name="stick">,
    @sex=#<Man:0x00007ff23f8454c0 @name="man">>,
   #<Wizard:0x00007ff23f845268
    @element=#<Thunder:0x00007ff23f845330 @name="thunder">,
    @equipment=#<Glove:0x00007ff23f8452e0 @name="glove">,
    @sex=#<Man:0x00007ff23f845380 @name="man">>,
   #<Wizard:0x00007ff23f8450d8
    @element=#<Wind:0x00007ff23f8451a0 @name="wind">,
    @equipment=#<Sword:0x00007ff23f845150 @name="sword">,
    @sex=#<Woman:0x00007ff23f8451f0 @name="woman">>,
   #<Wizard:0x00007ff23f844f98
    @element=#<Wind:0x00007ff23f845060 @name="wind">,
    @equipment=#<Stick:0x00007ff23f845010 @name="stick">,
    @sex=#<Woman:0x00007ff23f8450b0 @name="woman">>,
   #<Wizard:0x00007ff23f844e58
    @element=#<Wind:0x00007ff23f844f20 @name="wind">,
    @equipment=#<Glove:0x00007ff23f844ed0 @name="glove">,
    @sex=#<Woman:0x00007ff23f844f70 @name="woman">>,
   #<Wizard:0x00007ff23f844cf0
    @element=#<Water:0x00007ff23f844db8 @name="water">,
    @equipment=#<Sword:0x00007ff23f844d68 @name="sword">,
    @sex=#<Woman:0x00007ff23f844e08 @name="woman">>,
   #<Wizard:0x00007ff23f844bb0
    @element=#<Water:0x00007ff23f844c78 @name="water">,
    @equipment=#<Stick:0x00007ff23f844c28 @name="stick">,
    @sex=#<Woman:0x00007ff23f844cc8 @name="woman">>,
   #<Wizard:0x00007ff23f844a70
    @element=#<Water:0x00007ff23f844b38 @name="water">,
    @equipment=#<Glove:0x00007ff23f844ae8 @name="glove">,
    @sex=#<Woman:0x00007ff23f844b88 @name="woman">>,
   #<Wizard:0x00007ff23f844908
    @element=#<Thunder:0x00007ff23f8449d0 @name="thunder">,
    @equipment=#<Sword:0x00007ff23f844980 @name="sword">,
    @sex=#<Woman:0x00007ff23f844a20 @name="woman">>,
   #<Wizard:0x00007ff23f8447c8
    @element=#<Thunder:0x00007ff23f844890 @name="thunder">,
    @equipment=#<Stick:0x00007ff23f844840 @name="stick">,
    @sex=#<Woman:0x00007ff23f8448e0 @name="woman">>,
   #<Wizard:0x00007ff23f844688
    @element=#<Thunder:0x00007ff23f844750 @name="thunder">,
    @equipment=#<Glove:0x00007ff23f844700 @name="glove">,
    @sex=#<Woman:0x00007ff23f8447a0 @name="woman">>]>

- `#best_charactersメソッド`
[4] pry(main)> character_checker.best_characters
=> [#<Warrior:0x00007ff23f847108
  @element=#<Water:0x00007ff23f8471d0 @name="water">,
  @equipment=#<Sword:0x00007ff23f847180 @name="sword">,
  @sex=#<Man:0x00007ff23f847220 @name="man">>,
 #<Warrior:0x00007ff23f846528
  @element=#<Water:0x00007ff23f8465f0 @name="water">,
  @equipment=#<Sword:0x00007ff23f8465a0 @name="sword">,
  @sex=#<Woman:0x00007ff23f846640 @name="woman">>]

- `#best_attackersメソッド`
[5] pry(main)> character_checker.best_attackers
=> [#<Warrior:0x00007ff23f846d20
  @element=#<Thunder:0x00007ff23f846de8 @name="thunder">,
  @equipment=#<Sword:0x00007ff23f846d98 @name="sword">,
  @sex=#<Man:0x00007ff23f846e38 @name="man">>,
 #<Wizard:0x00007ff23f8453a8
  @element=#<Thunder:0x00007ff23f845470 @name="thunder">,
  @equipment=#<Stick:0x00007ff23f845420 @name="stick">,
  @sex=#<Man:0x00007ff23f8454c0 @name="man">>]

- `#best_defendersメソッド`
[6] pry(main)> character_checker.best_defenders
=> [#<Fighter:0x00007ff23f847ea0
  @element=#<Wind:0x00007ff23f847f68 @name="wind">,
  @equipment=#<Glove:0x00007ff23f847f18 @name="glove">,
  @sex=#<Woman:0x00007ff23e44c018 @name="woman">>]

### CODE REVIEW

File: power_checker.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  end
+
+  def strongest_characters
+    members.select { |character| character.attack + character.defence == highest_parameter }
+  end
+
+  private
+
+  def highest_parameter
+    members.map { |character| character.attack + character.defence }.max
```
Comment (kihaya) [inline power_checker.rb]: もしや
攻撃力100、防御力0
攻撃力50、防御力50
のキャラクターが同列ですか...？

**Review by kihaya — CHANGES_REQUESTED** [review]

**Comment by kihaya** [comment]
PRの説明がありませんね、pryでの実行結果などは貼らないのですか？

File: power_checker.rb (hunk 1 above)
Comment (yumayo14) [reply power_checker.rb]: はい、、そうなります、、、、

**Comment by yumayo14** [comment]
申し訳ありません、PRメッセージとコードで十分と考えて載せていませんでした。
確かに、pryでの実行結果などがないとどういうデータが返ってきているのかが分からないですね。
必要そうな箇所は、適宜載せる様にしたいと思います。

**Comment by kihaya** [comment]
このブランチは`Guild`クラスが含まれていますが、masterから生やしましたか？

**Comment by yumayo14** [comment]
スクリーンショット 2019-04-12 20 27 28

この箇所のことですか？

**Comment by kihaya** [comment]
プログラムの出力結果が見づらいので整形してください

**Comment by yumayo14** [comment]
分かりました、整形します。

**Comment by kihaya** [comment]
> 分かりました、整形します。

特に今回つくっているプログラムは単体で実行可能ではないので、
pry上でこういう入力をしたらこういう出力が返る、というのを以下のように示してほしいです。
```
 $ pry
[1] pry(main)> 1 + 1
=> 2
```

**Comment by kihaya** [comment]
`binding.pry`で止めてクラスの動作確認をしなくてもpryからファイルをロードしてテストをすれば良いと思うのです。

File: spec/character_checker_spec.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+describe CharacterChecker do
+  let(:all_guild_member) { Guild.new.member }
+  let(:character_checker) { CharacterChecker.new(all_guild_member) }
+  describe '#best_characters' do
+    it 'キャラクターの数は2人' do
+      expect(character_checker.best_characters.count).to eq 2
+    end
+    it '1人目は剣を装備した水属性の男性戦士' do
+      expect(character_checker.best_characters.first.class.name).to eq 'Warrior'
```
Comment (kihaya) [inline spec/character_checker_spec.rb]: ここだけなんで`class`の判定ではないのでしょう。

**Comment by yumayo14** [comment]
> binding.pryで止めてクラスの動作確認をしなくてもpryからファイルをロードしてテストをすれば良いと思うのです。

おっしゃる通りでした、修正します🙇‍♂️

File: spec/character_checker_spec.rb (hunk 2 above)
Comment (yumayo14) [reply spec/character_checker_spec.rb]: eq マッチャーで比較するときは、文字列か数値でないといけないと思い込んでいました。
classオブジェクトでも比較できたので修正します。

File: spec/character_checker_spec.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
+require_relative '../character_checker'
+require_relative '../guild'
+
+describe CharacterChecker do
+  let(:all_guild_member) { Guild.new.member }
+  let(:character_checker) { CharacterChecker.new(all_guild_member) }
+  describe '#best_characters' do
+    it 'キャラクターの数は2人' do
+      expect(character_checker.best_characters.count).to eq 2
```
Comment (kihaya) [inline spec/character_checker_spec.rb]: ここもsubjectにしたほうがいいですね。

**Review by kihaya — CHANGES_REQUESTED** [review]

**Comment by yumayo14** [comment]
こちらも書き直せたので、お手隙でお願いします。

**Comment by kihaya** [comment]
rebaseして向き先を`master`にしていただけると。

**Comment by yumayo14** [comment]
PRの向きを変更しました、よろしくお願いします🙇‍♂️

**Review by kihaya — APPROVED** [review]

**Pull request merged**